core/modules/replica/QservMgtServices.cc

```diff
@@ -0,0 +1,342 @@
+ */
+
+// Class header
+#include "replica/QservMgtServices.h"
+
+// System headers
+#include <stdexcept>
+
+// Third party headers
+#include "XrdSsi/XrdSsiProvider.hh"
+#include "XrdSsi/XrdSsiService.hh"
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "replica/Configuration.h"
+#include "replica/ServiceProvider.h"
+
+/// This C++ symbol is provided by the SSI shared library
+extern XrdSsiProvider* XrdSsiProviderClient;
+
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.QservMgtServices");
+
+} /// namespace
+
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+//////////////////////////////////////////////////////////////////////////////////
+//////////////////////////  QservMgtRequestWrapperImpl  //////////////////////////
+//////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Request-type specific wrappers
+ */
+template <class  T>
+struct QservMgtRequestWrapperImpl
+    :   QservMgtRequestWrapper {
+
+    /// The implementation of the vurtual method defined in the base class
+    virtual void notify() {
+        if (_onFinish == nullptr) { return; }
+        _onFinish(_request);
+    }
+
+    QservMgtRequestWrapperImpl(typename T::Ptr const& request,
+                               typename T::CallbackType onFinish)
+        :   QservMgtRequestWrapper(),
+            _request(request),
+            _onFinish(onFinish) {
+    }
+
+    /// Destructor
+    ~QservMgtRequestWrapperImpl() override = default;
+
+    /// Implement a virtual method of the base class
+    std::shared_ptr<QservMgtRequest> request() const override {
+        return _request;
+    }
+
+private:
+
+    // The context of the operation
+
+    typename T::Ptr       _request;
+    typename T::CallbackType _onFinish;
+};
+
+////////////////////////////////////////////////////////////////////////
+//////////////////////////  QservMgtServices  //////////////////////////
+////////////////////////////////////////////////////////////////////////
+
+QservMgtServices::Ptr QservMgtServices::create(ServiceProvider::Ptr const& serviceProvider) {
+    return QservMgtServices::Ptr(
+        new QservMgtServices(serviceProvider));
+}
+
+QservMgtServices::QservMgtServices(ServiceProvider::Ptr const& serviceProvider)
+    :   _serviceProvider(serviceProvider),
+        _io_service(),
+        _work(nullptr),
+        _registry() {
+}
+
+AddReplicaQservMgtRequest::Ptr QservMgtServices::addReplica(
+                                        unsigned int chunk,
+                                        std::vector<std::string> const& databases,
+                                        std::string const& worker,
+                                        AddReplicaQservMgtRequest::CallbackType onFinish,
+                                        std::string const& jobId,
+                                        unsigned int requestExpirationIvalSec) {
+
+    util::Lock lock(_mtx, "QservMgtServices::addReplica");
```

Comment:
you need to protect the map `_registry` from parallel access, but the lock need to be held for the rest of this function?

Reply:
Yes. There are two big reasons of that further in that method. First of all there is an asynchronous operation declared in line:

128:  AddReplicaQservMgtRequest::Ptr const request =
...
135:            [manager] (QservMgtRequest::Ptr const& request) {
136:                manager->finish(request->id());
137:           }

That "finish(...)" should not be called before the "addReplica()" method finishes registering the request in the registry at:
143:    _registry[request->id()] = ...

The method is very short. So, it's reasonable to keep the lock through the end of the method.

Reply:
If my reading is correct, `request` should be an inactive object until `request->start` at line 148 so it should be safe to 1. make it, 2. lock & insert in the map & unlock, 3. call start on it.

Reply:
That's true. However, I don't see any significant improvement in limiting a scope of the lock. Though, I understand your concern - a call to the request's method "start()" may block for a prolonged period of time thus preventing the service ("QservMgtServices") from service other requests until the call is over. Another potential implication of keeping a lock while "start()" is being called is that the request won't be able to initiate the immediate finalization (w/o throwing an exception during "start()") using the registered callback method. This is not the case now. However, someone may modify the code in the future to create a condition for a lock up here.
Having said that, I agree to limit a scope of the lock (in this and other similar methods of the device) :-)

Reply:
As Fritz pointed out I am strongly averse to holding a lock (or locks!) except for in a case where I can tell exactly what is being protected, and exactly why it is being protected. My bias will run towards that. You may have reasons to do it differently and that's fine, but I'm still gonna bring it up 😄

Reply:
You may be still looking at the outdated version the code. Twi issue has already been fixed.